File: backend/app/schemas/hotpost.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field, field_validator

from app.schemas.base import ORMModel


HotpostMode = Literal["trending", "rant", "opportunity"]
HotpostTimeFilter = Literal["week", "month", "year", "all"]


class HotpostSearchRequest(ORMModel):
    """爆帖速递搜索请求"""

    query: str = Field(min_length=2, max_length=2000)
    mode: HotpostMode | None = Field(default=None)
    subreddits: list[str] | None = Field(default=None)
    time_filter: HotpostTimeFilter | None = Field(default=None)
    limit: int = Field(default=30, ge=1, le=100)

    @field_validator("query")
    @classmethod
    def normalize_query(cls, value: str) -> str:
        cleaned = value.strip()
        if len(cleaned) < 2:
            raise ValueError("query must contain at least 2 non-whitespace characters")
        return cleaned

    @field_validator("mode", mode="before")
    @classmethod
    def normalize_mode(cls, value: object) -> str | None:
        if value is None:
            return None
        cleaned = str(value).strip().lower()
        return cleaned or None

    @field_validator("time_filter", mode="before")
    @classmethod
    def normalize_time_filter(cls, value: object) -> str | None:
        if value is None:
            return None
        cleaned = str(value).strip().lower()
        return cleaned or None

    @field_validator("subreddits", mode="before")
    @classmethod
    def normalize_subreddits(cls, value: object) -> list[str] | None:
        if value is None:
            return None
        items = []
        for raw in list(value):
            name = str(raw).strip().lower()
            if not name:
                continue
            if not name.startswith("r/"):
                name = f"r/{name}"
            items.append(name)
        if not items:
            return None
        if len(items) > 10:
            raise ValueError("subreddits最多10个")
        return items


class HotpostComment(ORMModel):
    comment_fullname: str | None = None
    author: str | None = None
    body: str | None = None
    score: int | None = None
    permalink: str | None = None


class Hotpost(ORMModel):
    rank: int
    id: str
    title: str
    body_preview: str | None = None
    score: int
    num_comments: int
    heat_score: int | None = None
    rant_score: float | None = None
    rant_signals: list[str] = Field(default_factory=list)
    resonance_count: int | None = None
    subreddit: str
    author: str | None = None
    reddit_url: str
    created_utc: float
    signals: list[str]
    signal_score: float
    why_relevant: str | None = None
    why_important: str | None = None
    top_comments: list[HotpostComment] = Field(default_factory=list)


class HotpostTopicEvidence(ORMModel):
    title: str
    score: int | None = None
    comments: int | None = None
    subreddit: str | None = None
    url: str | None = None
    key_quote: str | None = None


class HotpostTopic(ORMModel):
    rank: int
    topic: str
    heat_score: int | None = None
    time_trend: str | None = None
    key_takeaway: str | None = None
    evidence: list[HotpostTopicEvidence] = Field(default_factory=list)
    evidence_posts: list["Hotpost"] = Field(default_factory=list)


class PainPoint(ORMModel):
    category: str
    category_en: str | None = None
    description: str | None = None
    mentions: int | None = None
    percentage: float | None = None
    key_takeaway: str | None = None
    severity: str | None = None
    sample_quotes: list[str] = Field(default_factory=list)
    evidence_urls: list[str] = Field(default_factory=list)
    rank: int | None = None
    user_voice: str | None = None
    business_implication: str | None = None
    evidence: list[HotpostTopicEvidence] = Field(default_factory=list)
    evidence_posts: list[Hotpost] = Field(default_factory=list)


class CompetitorMention(ORMModel):
    name: str
    sentiment: str | None = None
    mentions: int | None = None
    sentiment_score: float | None = None
    common_praise: list[str] = Field(default_factory=list)
    common_complaint: list[str] = Field(default_factory=list)
    typical_context: str | None = None
    sample_quote: str | None = None
    vs_adobe: str | None = None
    evidence_quote: str | None = None


class MigrationIntent(ORMModel):
    already_left: str | None = None
    considering: str | None = None
    staying_reluctantly: str | None = None
    total_mentions: int | None = None
    percentage: float | None = None
    status_distribution: dict[str, float] | None = None
    destinations: list[dict[str, object]] = Field(default_factory=list)
    key_quote: str | None = None


class TopQuote(ORMModel):
    quote: str
    score: int | None = None
    subreddit: str | None = None
    url: str | None = None


class CurrentWorkaround(ORMModel):
    solution: str | None = None
    pain: str | None = None
    name: str | None = None
    satisfaction: str | None = None


class UnmetNeedEvidence(ORMModel):
    title: str
    score: int | None = None
    comments: int | None = None
    me_too_comments: int | None = None
    subreddit: str | None = None
    url: str | None = None
    key_quote: str | None = None


class UnmetNeed(ORMModel):
    rank: int | None = None
    need: str
    need_en: str | None = None
    urgency: str | None = None
    mentions: int | None = None
    demand_signal: str | None = None
    me_too_count: int | None = None
    willingness_to_pay: str | None = None
    pay_evidence: str | None = None
    price_range: str | None = None
    key_takeaway: str | None = None
    user_voice: str | None = None
    current_workarounds: list[CurrentWorkaround] = Field(default_factory=list)
    opportunity_insight: str | None = None
    evidence: list[UnmetNeedEvidence] = Field(default_factory=list)
    evidence_posts: list[Hotpost] = Field(default_factory=list)


class ExistingTool(ORMModel):
    name: str
    sentiment: str | None = None
    mentions: int | None = None
    sentiment_score: float | None = None
    common_praise: list[str] = Field(default_factory=list)
    common_complaint: list[str] = Field(default_factory=list)
    praised_for: list[str] = Field(default_factory=list)
    criticized_for: list[str] = Field(default_factory=list)
    gap_analysis: str | None = None


class UserSegment(ORMModel):
    segment: str
    percentage: str | None = None
    core_need: str | None = None
    key_need: str | None = None
    price_sensitivity: str | None = None
    typical_quote: str | None = None


class MarketOpportunity(ORMModel):
    gap: str | None = None
    target_user: str | None = None
    pricing_hint: str | None = None
    gtm_channel: str | None = None
    strength: str | None = None
    unmet_gap: str | None = None
    demand_signal: str | None = None
    competition_level: str | None = None
    recommendation: str | None = None


class NextSteps(ORMModel):
    deepdive_available: bool | None = None
    deepdive_token: str | None = None
    suggested_keywords: list[str] = Field(default_factory=list)


class HotpostSearchResponse(ORMModel):
    query_id: str
    query: str
    mode: HotpostMode
    search_time: datetime
    from_cache: bool
    status: str | None = None
    position: int | None = None
    estimated_wait_seconds: int | None = None
    summary: str
    markdown_report: str | None = None
    top_posts: list[Hotpost]
    key_comments: list[HotpostComment]
    top_rants: list[Hotpost] | None = None
    top_discovery_posts: list[Hotpost] | None = None
    pain_points: list[PainPoint] | None = None
    opportunities: list[dict[str, Any]] | None = None
    trending_keywords: list[str] | None = None
    topics: list[HotpostTopic] | None = None
    competitor_mentions: list[CompetitorMention] | None = None
    migration_intent: MigrationIntent | None = None
    top_quotes: list[TopQuote] | None = None
    unmet_needs: list[UnmetNeed] | None = None
    existing_tools: list[ExistingTool] | None = None
    user_segments: list[UserSegment] | None = None
    market_opportunity: MarketOpportunity | None = None
    communities: list[str]
    related_queries: list[str]
    evidence_count: int
    community_distribution: dict[str, str]
    sentiment_overview: dict[str, float]
    rant_intensity: dict[str, float] | None = None
    need_urgency: dict[str, float] | None = None
    opportunity_strength: str | None = None
    confidence: str
    reliability_note: str | None = None
    next_steps: NextSteps | None = None
    notes: list[str] | None = None
    debug_info: dict[str, Any] | None = None


class HotpostDeepdiveResponse(ORMModel):
    deepdive_token: str
    expires_in_seconds: int


class HotpostDeepdiveRequest(ORMModel):
    query_id: str
    product_desc: str | None = None
    seed_subreddits: list[str] | None = None


__all__ = [
    "HotpostMode",
    "HotpostTimeFilter",
    "HotpostSearchRequest",
    "HotpostSearchResponse",
    "HotpostDeepdiveResponse",
    "HotpostDeepdiveRequest",
    "Hotpost",
    "HotpostComment",
    "HotpostTopic",
    "HotpostTopicEvidence",
    "PainPoint",
    "CompetitorMention",
    "MigrationIntent",
    "TopQuote",
    "UnmetNeed",
    "UnmetNeedEvidence",
    "CurrentWorkaround",
    "ExistingTool",
    "UserSegment",
    "MarketOpportunity",
    "NextSteps",
]

File: frontend/src/types/hotpost.ts
// frontend/src/types/hotpost.ts

export type HotPostMode = 'trending' | 'rant' | 'opportunity';
export type TimeFilter = 'week' | 'month' | 'year' | 'all';
export type ConfidenceLevel = 'high' | 'medium' | 'low' | 'none';

export interface HotPostRequest {
  query: string;
  mode?: HotPostMode;
  subreddits?: string[]; // Max 10
  time_filter?: TimeFilter;
  limit?: number; // 1-100, default 30
}

export interface HotPost {
  rank: number;
  id: string;
  title: string;
  body_preview: string;
  score: number;
  num_comments: number;
  heat_score?: number;
  rant_score?: number;
  rant_signals?: string[];
  resonance_count?: number;
  subreddit: string;
  author: string;
  reddit_url: string;
  created_utc: number;
  signals: string[];
  signal_score: number;
  why_relevant?: string;
  why_important?: string;
  top_comments: Array<{
    author: string;
    body: string;
    score: number;
    id?: string;
    comment_fullname?: string;
    permalink?: string;
  }>;
}

// --- Rant Mode Structures ---

export interface PainPoint {
  rank?: number;
  category: string;
  category_en?: string;
  description?: string;
  severity?: 'critical' | 'high' | 'medium' | 'low' | string;
  mentions?: number;
  percentage?: number;
  key_takeaway?: string;
  user_voice?: string;
  business_implication?: string;
  sample_quotes?: string[];
  evidence_urls?: string[];
  evidence?: Array<{
    title?: string;
    score?: number;
    comments?: number;
    url?: string;
    key_quote?: string;
  }>;
  evidence_posts?: HotPost[];
}

export interface CompetitorMention {
  name: string;
  sentiment: 'positive' | 'neutral' | 'negative';
  mentions: number;
  sentiment_score?: number;
  common_praise?: string[];
  common_complaint?: string[];
  typical_context?: string;
  sample_quote?: string;
  vs_adobe?: string;
  evidence_quote?: string;
}

export interface MigrationIntent {
  already_left?: string; // e.g. "20%"
  considering?: string;
  staying_reluctantly?: string;
  total_mentions?: number;
  percentage?: number;
  status_distribution?: Record<string, number>;
  destinations?: Array<{ name?: string; mentions?: number; sentiment?: string }>;
  key_quote?: string;
}

// --- Opportunity Mode Structures ---

export interface Opportunity {
  rank?: number;
  need?: string; // The unmet need
  need_en?: string;
  urgency?: string;
  mentions?: number;
  summary?: string;
  mentions?: number;
  me_too_count?: number; // demand_signal
  willingness_to_pay?: 'high' | 'medium' | 'low';
  pay_evidence?: string;
  price_range?: string;
  key_takeaway?: string;
  user_voice?: string;
  current_workarounds?: Array<{ solution?: string; pain?: string; name?: string; satisfaction?: string }>;
  opportunity_signal?: 'high' | 'medium' | 'low';
  evidence?: Array<{
    title?: string;
    score?: number;
    comments?: number;
    me_too_comments?: number;
    subreddit?: string;
    url?: string;
    key_quote?: string;
  }>;
  evidence_posts?: HotPost[];
}

export interface ExistingTool {
    name: string;
    sentiment: 'positive' | 'neutral' | 'negative';
    mentions?: number;
    sentiment_score?: number;
    common_praise?: string[];
    common_complaint?: string[];
    praised_for?: string[];
    criticized_for?: string[];
    pros?: string[]; // legacy alias
    cons?: string[]; // legacy alias
    gap_analysis?: string;
}

export interface UserSegment {
    segment?: string;
    percentage?: string;
    core_need?: string;
    key_need?: string;
    price_sensitivity?: string;
    segment_name?: string; // legacy alias
    needs?: string[]; // legacy alias
    typical_profile?: string; // legacy alias
    typical_quote?: string;
}

export interface MarketOpportunity {
  gap?: string;
  target_user?: string;
  pricing_hint?: string;
  gtm_channel?: string;
  strength?: string;
  unmet_gap?: string;
  demand_signal?: string;
  competition_level?: string;
  recommendation?: string;
}

// --- Trending Mode Structures ---

export interface TrendingTopic {
    rank: number;
    topic: string;
    heat_score: number;
    time_trend: '新兴🆕' | '持续热门' | '下降中↓' | string;
    key_takeaway: string;
    evidence: Array<{
        title: string;
        score: number;
        comments?: number;
        subreddit: string;
        url: string;
        key_quote?: string;
    }>;
    evidence_posts?: HotPost[];
}

// --- Main Response ---

export interface HotPostResponse {
  query_id: string;
  query: string;
  mode: HotPostMode;
  search_time: string;
  from_cache: boolean;
  status?: 'queued' | 'waiting' | 'processing' | 'completed' | 'success' | 'failed';
  
  // Queue Info
  position?: number;
  estimated_wait_seconds?: number;

  // Summary & Stats
  summary: string;
  markdown_report?: string;
  confidence: ConfidenceLevel;
  evidence_count: number;
  community_distribution: Record<string, number | string>;
  sentiment_overview?: {
    positive: number;
    neutral: number;
    negative: number;
  };
  rant_intensity?: Record<string, number>;
  need_urgency?: Record<string, number>;
  opportunity_strength?: string;

  // Core Data
  top_posts: HotPost[];
  key_comments?: any[];
  top_rants?: HotPost[];
  top_discovery_posts?: HotPost[];

  // Mode Specific - Rant
  pain_points?: PainPoint[];
  competitor_mentions?: CompetitorMention[];
  migration_intent?: MigrationIntent;

  // Mode Specific - Opportunity
  opportunities?: Opportunity[];
  unmet_needs?: Opportunity[]; // Backend might send 'unmet_needs', let's support both or alias
  existing_tools?: ExistingTool[];
  user_segments?: UserSegment[];
  market_opportunity?: MarketOpportunity | string;

  // Mode Specific - Trending
  trending_keywords?: string[];
  topics?: TrendingTopic[]; // Backend 'topics'

  reliability_note?: string;
  next_steps?: {
    deepdive_available?: boolean;
    deepdive_token?: string | null;
    suggested_keywords?: string[];
  };
  notes?: string[];
  debug_info?: {
    query_source?: string;
    search_query?: string;
    query_parts?: string[];
    keywords?: string[];
    time_filter?: string;
    sort?: string;
    subreddits?: string[];
    raw_posts?: number;
    filtered_posts?: number;
    relevance_filtered?: number;
  };

  // Meta
  communities: Array<string | {
    name: string;
    subscribers?: number;
    active_users?: number;
    description?: string;
  }>;
  related_queries: string[];
}

export interface DeepDiveRequest {
  query_id: string;
  product_desc?: string;
  seed_subreddits?: string[];
}

export interface DeepDiveResponse {
  deepdive_token: string;
  expires_in_seconds: number;
}
